### PR TITLE
Support arbitrary indexes

### PR DESCRIPTION
**Pull request by kuba-orlik**
With this change, field-types can tell Sealious to create an index in the database, and specify the index type. That way, one will be able to leverage the advantages e.g. of MongoDB's [2dsphere index](https://docs.mongodb.com/manual/core/2dsphere/) :)

### CODE REVIEW

File: lib/base-chips/field-types/text.js [hunk 1]
```diff
@@ -4,12 +4,10 @@ const escape = require("escape-html");
 
 module.exports = {
 	name: "text",
-	full_text_search_enabled: function(params){
-		if (params && params.include_in_search){
-			return true;
-		} else if (params && params.max_length > 200){
-			return true;
-		} else {
+	has_index: function(params){
```
Comment (adwydman) [inline lib/base-chips/field-types/text.js]: Why did you replace `full_text_search_enabled` with `has_index`?

File: lib/chip-types/field.js [hunk 2]
```diff
@@ -17,7 +17,7 @@ function Field (declaration){
 		this[method_name] = (function(method_name){
 			return function(){
 				const arguments_array = Object.keys(arguments).map((key)=>arguments[key]);
-				arguments_array.splice(1, 0, this.params);
+				arguments_array.splice(1, 0, self.params);
```
Comment (adwydman) [inline lib/chip-types/field.js]: One can wonder how did this even work :D

File: lib/base-chips/field-types/text.js (hunk 1 above)
Comment (kuba-orlik) [inline lib/base-chips/field-types/text.js]: That's the whole deal with this change. `full_text_search` supported only `text` indexes. Now we can support arbitrary indexes, like `sphere2d`. From user-reference:

> has_index: optional. Whether or not to instruct the Datastore to create an index on that field. In order to include the contents of the fields of this type in full-text-search, return "text" here.

File: lib/chip-types/field.js (hunk 2 above)
Comment (kuba-orlik) [inline lib/chip-types/field.js]: I agree, it's kind of mysterious. It works cool, but we have to refactor it sometime

File: lib/base-chips/field-types/text.js (hunk 1 above)
Comment (adwydman) [inline lib/base-chips/field-types/text.js]: Please explain what do you mean by `index`?

Comment (kuba-orlik) [inline lib/base-chips/field-types/text.js]: In a nutshell, 'index' in a database is a data structure that speeds up certain queries. Without indexing, when you want to find all items matching a specified query, the db engine would have to look through each entry in the collection. With indexes, some additional metadata is stored to speed up the process, so the database knows for example where *not* to look :) Both SQL and noSQL databases have some sort of support of indexes: 

https://docs.mongodb.com/manual/indexes/
https://www.postgresql.org/docs/current/static/sql-createindex.html

Comment (kuba-orlik) [inline lib/base-chips/field-types/text.js]: PS you might want to read the corresponding changes in the sealious-datastore-dbs-common-parts https://github.com/sealcode/sealious-datastore-dbs-common-part/commit/649b7446f9f28f535339390747ad8b481dcba8ab

**Pull request merged**